octopus/src/main/scala/octopus/AsyncValidator.scala

```diff
@@ -1,31 +1,30 @@
 package octopus
 
-import scala.concurrent.{ExecutionContext, Future}
+import scala.language.higherKinds
 
-trait AsyncValidator[T] {
-
-  def validate(obj: T)
-              (implicit ec: ExecutionContext): Future[List[ValidationError]]
+trait AsyncValidator[M[_], T] {
+  def validate(obj: T)(implicit app: App[M]): M[List[ValidationError]]
 }
 
 object AsyncValidator {
 
-  def instance[T](f: (T, ExecutionContext) => Future[List[ValidationError]]): AsyncValidator[T] =
-    new AsyncValidator[T] {
-      def validate(obj: T)(implicit ec: ExecutionContext): Future[List[ValidationError]] =
-        f(obj, ec)
+  def instance[M[_], T](f: T => M[List[ValidationError]]): AsyncValidator[M, T] = {
+    new AsyncValidator[M, T] {
+      def validate(obj: T)(implicit app: App[M]): M[List[ValidationError]] =
+        f(obj)
     }
+  }
 
-  def lift[T](v: Validator[T]): AsyncValidator[T] =
-    instance { (obj: T, _: ExecutionContext) =>
-      Future.successful(v.validate(obj))
+  def lift[M[_]: App, T](v: Validator[T]): AsyncValidator[M, T] =
+    instance { obj =>
+      App[M].pure(v.validate(obj))
     }
 
-  def apply[T]: AsyncValidator[T] =
+  def apply[M[_]: App, T]: AsyncValidator[M, T] =
```

Comment:
Breaking change here. WDYT about `AsyncValidator[T]` and just smart constructor like `AsyncValidatorM[Task, T]`?

For clarity: 2-arg type class is perfectly fine, what I refer to is only a way of constructing default instances which is part of the DSL.

Reply:
I think that this problem is a way more significant then we think.
How should we treat all instances returned by `AsyncValidatorAsyncOps` methods? all of them are exposed to end user.

Maybe we should double them all? 
That might be the pain we have to go through because of early `Future` commitment in the user interface.

The second option is to rename monad to AsyncValidatorM and provide user only this alias(in case they already used `AsyncValiator` name in their code)

`type AsyncValidator[T] = AsyncValidatorM[Future, T]`